File: tests/test_gcPCA.R
# Tests for generalized contrastive PCA (R implementation)
#
# Run with: Rscript tests/test_gcPCA.R
# from the repository root directory.

# Source the R implementations
source("R/gcPCA.R")
source("R/sparse_gcPCA.R")

# ---- Test helpers ----

test_count <- 0
pass_count <- 0
fail_count <- 0
failures <- list()

assert <- function(condition, msg) {
  test_count <<- test_count + 1
  if (isTRUE(condition)) {
    pass_count <<- pass_count + 1
    cat(sprintf("  PASS: %s\n", msg))
  } else {
    fail_count <<- fail_count + 1
    failures[[length(failures) + 1]] <<- msg
    cat(sprintf("  FAIL: %s\n", msg))
  }
}

assert_close <- function(x, y, tol = 1e-6, msg = "") {
  assert(all(abs(x - y) < tol), msg)
}

# ---- Generate synthetic data ----

set.seed(42)
n_a <- 80
n_b <- 60
p <- 20

Rb <- matrix(rnorm(n_b * p), nrow = n_b, ncol = p)
Ra <- matrix(rnorm(n_a * p), nrow = n_a, ncol = p)
Ra[, 1] <- Ra[, 1] + rnorm(n_a) * 5
Ra[, 2] <- Ra[, 2] + rnorm(n_a) * 3

# Small data for quick tests
set.seed(42)
n_a_sm <- 30
n_b_sm <- 25
p_sm <- 8
Rb_sm <- matrix(rnorm(n_b_sm * p_sm), nrow = n_b_sm, ncol = p_sm)
Ra_sm <- matrix(rnorm(n_a_sm * p_sm), nrow = n_a_sm, ncol = p_sm)
Ra_sm[, 1] <- Ra_sm[, 1] + rnorm(n_a_sm) * 4

cat("\n========================================\n")
cat("  R gcPCA Tests\n")
cat("========================================\n\n")

# ---- Test gcPCA: non-orthogonal methods ----

cat("--- gcPCA non-orthogonal methods ---\n")

for (method in c("v1", "v2", "v3", "v4")) {
  result <- tryCatch({
    mdl <- gcPCA(Ra, Rb, method = method)
    TRUE
  }, error = function(e) {
    cat(sprintf("  ERROR in %s: %s\n", method, e$message))
    FALSE
  })
  assert(result, sprintf("gcPCA %s fits without error", method))

  if (result) {
    mdl <- gcPCA(Ra, Rb, method = method)
    assert(nrow(mdl$loadings) == p, sprintf("gcPCA %s: loadings has %d rows (features)", method, p))
    assert(!is.null(mdl$Ra_scores), sprintf("gcPCA %s: Ra_scores exists", method))
    assert(!is.null(mdl$Rb_scores), sprintf("gcPCA %s: Rb_scores exists", method))
    assert(!is.null(mdl$objective_values), sprintf("gcPCA %s: objective_values exists", method))
  }
}

# ---- Test gcPCA: orthogonal methods ----

cat("\n--- gcPCA orthogonal methods ---\n")

for (method in c("v2.1", "v3.1", "v4.1")) {
  result <- tryCatch({
    mdl <- gcPCA(Ra_sm, Rb_sm, method = method, Ncalc = 4)
    TRUE
  }, error = function(e) {
    cat(sprintf("  ERROR in %s: %s\n", method, e$message))
    FALSE
  })
  assert(result, sprintf("gcPCA %s fits without error", method))

  if (result) {
    mdl <- gcPCA(Ra_sm, Rb_sm, method = method, Ncalc = 4)
    X <- mdl$loadings
    gram <- t(X) %*% X
    off_diag_max <- max(abs(gram - diag(diag(gram))))
    assert(off_diag_max < 1e-5, sprintf("gcPCA %s: loadings are orthogonal (max off-diag = %.2e)", method, off_diag_max))
  }
}

# ---- Test gcPCA: v4 objective values bounded ----

cat("\n--- gcPCA v4 value bounds ---\n")

mdl <- gcPCA(Ra, Rb, method = "v4")
assert(all(mdl$objective_values >= -1 - 1e-10), "v4 objective values >= -1")
assert(all(mdl$objective_values <= 1 + 1e-10), "v4 objective values <= 1")

# ---- Test gcPCA: v2 objective values positive ----

cat("\n--- gcPCA v2 value bounds ---\n")

mdl <- gcPCA(Ra, Rb, method = "v2")
assert(all(mdl$objective_values > 0), "v2 objective values > 0")

# ---- Test gcPCA: different feature counts should error ----

cat("\n--- gcPCA input validation ---\n")

result <- tryCatch({
  gcPCA(matrix(rnorm(50), 10, 5), matrix(rnorm(60), 10, 6), method = "v4")
  FALSE
}, error = function(e) TRUE)
assert(result, "Different feature counts raises error")

# ---- Test gcPCA: invalid method should error ----

result <- tryCatch({
  gcPCA(Ra, Rb, method = "v5")
  FALSE
}, error = function(e) TRUE)
assert(result, "Invalid method raises error")

# ---- Test gcPCA: rank-deficient data ----

cat("\n--- gcPCA rank-deficient data ---\n")

set.seed(42)
Ra_rd <- matrix(rnorm(5 * 20), nrow = 5, ncol = 20)
Rb_rd <- matrix(rnorm(5 * 20), nrow = 5, ncol = 20)
result <- tryCatch({
  mdl <- suppressWarnings(gcPCA(Ra_rd, Rb_rd, method = "v4"))
  !is.null(mdl$loadings)
}, error = function(e) {
  cat(sprintf("  ERROR: %s\n", e$message))
  FALSE
})
assert(result, "Rank-deficient data handled")

# ---- Test gcPCA: equal data gives zero v4 values ----

cat("\n--- gcPCA equal data ---\n")

set.seed(42)
data_eq <- matrix(rnorm(30 * 10), nrow = 30, ncol = 10)
mdl <- gcPCA(data_eq, data_eq, method = "v4", normalize_flag = FALSE)
assert_close(mdl$objective_values, rep(0, length(mdl$objective_values)), tol = 1e-10,
             msg = "Equal data gives zero v4 objective values")

# ---- Test gcPCA: predict method ----

cat("\n--- gcPCA predict ---\n")

mdl <- gcPCA(Ra_sm, Rb_sm, method = "v4")

# predict with no new data returns training scores
pred_train <- predict(mdl)
assert(!is.null(pred_train$Ra_scores), "predict() with no args returns Ra_scores")
assert(!is.null(pred_train$Rb_scores), "predict() with no args returns Rb_scores")

# predict with new data
set.seed(99)
Ra_new <- matrix(rnorm(10 * p_sm), nrow = 10, ncol = p_sm)
pred_new <- predict(mdl, Ra = Ra_new)
assert(nrow(pred_new$Ra_scores) == 10, "predict() with new Ra has correct rows")

# ---- Test gcPCA: reproducibility ----

cat("\n--- gcPCA reproducibility ---\n")

mdl1 <- gcPCA(Ra, Rb, method = "v4")
mdl2 <- gcPCA(Ra, Rb, method = "v4")
assert_close(abs(mdl1$loadings), abs(mdl2$loadings), tol = 1e-10,
             msg = "Same data gives same loadings")

# ---- Test sparse_gcPCA: basic fit ----

cat("\n--- sparse_gcPCA ---\n")

lambdas <- c(0.1, 0.5, 1.0)

for (method in c("v2", "v3", "v4")) {
  result <- tryCatch({
    mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = method, Nsparse = 2, lasso_penalty = lambdas)
    TRUE
  }, error = function(e) {
    cat(sprintf("  ERROR in sparse %s: %s\n", method, e$message))
    FALSE
  })
  assert(result, sprintf("sparse_gcPCA %s fits without error", method))

  if (result) {
    mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = method, Nsparse = 2, lasso_penalty = lambdas)
    assert(length(mdl$sparse_loadings) == length(lambdas),
           sprintf("sparse_gcPCA %s: correct number of loading sets", method))
    assert(nrow(mdl$sparse_loadings[[1]]) == p_sm,
           sprintf("sparse_gcPCA %s: loadings have correct number of rows", method))
    assert(ncol(mdl$sparse_loadings[[1]]) == 2,
           sprintf("sparse_gcPCA %s: loadings have Nsparse columns", method))
  }
}

# ---- Test sparse_gcPCA v1: basic fit ----

cat("\n--- sparse_gcPCA v1 ---\n")

result <- tryCatch({
  mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = "v1", Nsparse = 2, lasso_penalty = lambdas)
  TRUE
}, error = function(e) {
  cat(sprintf("  ERROR in sparse v1: %s\n", e$message))
  FALSE
})
assert(result, "sparse_gcPCA v1 fits without error")

if (result) {
  mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = "v1", Nsparse = 2, lasso_penalty = lambdas)
  assert(length(mdl$sparse_loadings) == length(lambdas),
         "sparse_gcPCA v1: correct number of loading sets")
  assert(nrow(mdl$sparse_loadings[[1]]) == p_sm,
         "sparse_gcPCA v1: loadings have correct number of rows")
}

# ---- Test sparse_gcPCA v1: scores output ----

cat("\n--- sparse_gcPCA v1 scores ---\n")

mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = "v1", Nsparse = 2, lasso_penalty = lambdas)
assert(length(mdl$Ra_scores) == length(lambdas), "v1 Ra_scores list length matches lambdas")
assert(length(mdl$Rb_scores) == length(lambdas), "v1 Rb_scores list length matches lambdas")
assert(length(mdl$Ra_values) == length(lambdas), "v1 Ra_values list length matches lambdas")
assert(length(mdl$Rb_values) == length(lambdas), "v1 Rb_values list length matches lambdas")

# ---- Test sparse_gcPCA: scores output ----

cat("\n--- sparse_gcPCA scores ---\n")

mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = "v4", Nsparse = 2, lasso_penalty = lambdas)
assert(length(mdl$Ra_scores) == length(lambdas), "Ra_scores list length matches lambdas")
assert(length(mdl$Rb_scores) == length(lambdas), "Rb_scores list length matches lambdas")
assert(length(mdl$Ra_values) == length(lambdas), "Ra_values list length matches lambdas")
assert(length(mdl$Rb_values) == length(lambdas), "Rb_values list length matches lambdas")

# ---- Test sparse_gcPCA: predict method ----

cat("\n--- sparse_gcPCA predict ---\n")

mdl <- sparse_gcPCA(Ra_sm, Rb_sm, method = "v4", Nsparse = 2, lasso_penalty = c(0.1))

# predict with no new data
pred_train <- predict(mdl)
assert(!is.null(pred_train$Ra_transform), "sparse predict() with no args returns Ra_transform")
assert(!is.null(pred_train$Rb_transform), "sparse predict() with no args returns Rb_transform")

# predict with new data
set.seed(99)
Ra_new <- matrix(rnorm(10 * p_sm), nrow = 10, ncol = p_sm)
pred_new <- predict(mdl, Ra = Ra_new)
assert(!is.null(pred_new$Ra_transform), "sparse predict() with new Ra works")

# ---- Summary ----

cat("\n========================================\n")
cat(sprintf("  Results: %d passed, %d failed out of %d tests\n", pass_count, fail_count, test_count))
cat("========================================\n")

if (fail_count > 0) {
  cat("\nFailed tests:\n")
  for (f in failures) {
    cat(sprintf("  - %s\n", f))
  }
  quit(status = 1)
} else {
  cat("\nAll tests passed!\n")
  quit(status = 0)
}

File: tests/test_gcPCA.m
% Tests for generalized contrastive PCA (MATLAB implementation)
%
% Run from repository root:
%   matlab -batch "cd tests; test_gcPCA"

% Add matlab source to path
addpath(fullfile(fileparts(mfilename('fullpath')), '..', 'matlab'));

test_count = 0;
pass_count = 0;
fail_count = 0;
failures = {};

%% Generate synthetic data
rng(42);
n_a = 80;
n_b = 60;
p = 20;

Rb = randn(n_b, p);
Ra = randn(n_a, p);
Ra(:, 1) = Ra(:, 1) + randn(n_a, 1) * 5;
Ra(:, 2) = Ra(:, 2) + randn(n_a, 1) * 3;

% Small data for quick tests
rng(42);
n_a_sm = 30;
n_b_sm = 25;
p_sm = 8;
Rb_sm = randn(n_b_sm, p_sm);
Ra_sm = randn(n_a_sm, p_sm);
Ra_sm(:, 1) = Ra_sm(:, 1) + randn(n_a_sm, 1) * 4;

fprintf('\n========================================\n');
fprintf('  MATLAB gcPCA Tests\n');
fprintf('========================================\n\n');

%% Test gcPCA: non-orthogonal methods
fprintf('--- gcPCA non-orthogonal methods ---\n');

for ver = [2, 3, 4]
    try
        [B, S, X] = gcPCA(Ra, Rb, ver);
        [test_count, pass_count] = check(true, sprintf('gcPCA v%d fits without error', ver), test_count, pass_count);

        [test_count, pass_count] = check(size(X, 1) == p, ...
            sprintf('gcPCA v%d: loadings has %d rows (features)', ver, p), test_count, pass_count);
        [test_count, pass_count] = check(~isempty(B.a), ...
            sprintf('gcPCA v%d: B.a exists', ver), test_count, pass_count);
        [test_count, pass_count] = check(~isempty(B.b), ...
            sprintf('gcPCA v%d: B.b exists', ver), test_count, pass_count);
        [test_count, pass_count] = check(~isempty(S.objval), ...
            sprintf('gcPCA v%d: S.objval exists', ver), test_count, pass_count);
    catch ME
        test_count = test_count + 1;
        fail_count = fail_count + 1;
        failures{end+1} = sprintf('gcPCA v%d: %s', ver, ME.message);
        fprintf('  FAIL: gcPCA v%d: %s\n', ver, ME.message);
    end
end

%% Test gcPCA v1 (contrastive PCA)
fprintf('\n--- gcPCA v1 (contrastive PCA) ---\n');
try
    [B, S, X] = gcPCA(Ra, Rb, 1);
    [test_count, pass_count] = check(true, 'gcPCA v1 fits without error', test_count, pass_count);
    [test_count, pass_count] = check(size(X, 1) == p, ...
        'gcPCA v1: loadings has correct rows', test_count, pass_count);
catch ME
    test_count = test_count + 1;
    fail_count = fail_count + 1;
    failures{end+1} = sprintf('gcPCA v1: %s', ME.message);
    fprintf('  FAIL: gcPCA v1: %s\n', ME.message);
end

%% Test gcPCA: orthogonal methods
fprintf('\n--- gcPCA orthogonal methods ---\n');

for ver = [2.1, 3.1, 4.1]
    try
        [B, S, X] = gcPCA(Ra_sm, Rb_sm, ver, 'Ncalc', 4);
        [test_count, pass_count] = check(true, sprintf('gcPCA v%.1f fits without error', ver), test_count, pass_count);

        gram = X' * X;
        off_diag = gram - diag(diag(gram));
        off_diag_max = max(abs(off_diag(:)));
        [test_count, pass_count] = check(off_diag_max < 1e-5, ...
            sprintf('gcPCA v%.1f: loadings orthogonal (max off-diag = %.2e)', ver, off_diag_max), test_count, pass_count);
    catch ME
        test_count = test_count + 1;
        fail_count = fail_count + 1;
        failures{end+1} = sprintf('gcPCA v%.1f: %s', ver, ME.message);
        fprintf('  FAIL: gcPCA v%.1f: %s\n', ver, ME.message);
    end
end

%% Test gcPCA: v4 objective values bounded
fprintf('\n--- gcPCA v4 value bounds ---\n');
[~, S, ~] = gcPCA(Ra, Rb, 4);
objval = S.objval;
[test_count, pass_count] = check(all(objval >= -1 - 1e-10), ...
    'v4 objective values >= -1', test_count, pass_count);
[test_count, pass_count] = check(all(objval <= 1 + 1e-10), ...
    'v4 objective values <= 1', test_count, pass_count);

%% Test gcPCA: v2 objective values positive
fprintf('\n--- gcPCA v2 value bounds ---\n');
[~, S, ~] = gcPCA(Ra, Rb, 2);
[test_count, pass_count] = check(all(S.objval > 0), ...
    'v2 objective values > 0', test_count, pass_count);

%% Test gcPCA: different feature counts should error
fprintf('\n--- gcPCA input validation ---\n');
try
    gcPCA(randn(10, 5), randn(10, 6), 4);
    test_count = test_count + 1;
    fail_count = fail_count + 1;
    failures{end+1} = 'Different feature counts should error';
    fprintf('  FAIL: Different feature counts should error\n');
catch
    [test_count, pass_count] = check(true, 'Different feature counts raises error', test_count, pass_count);
end

%% Test gcPCA: rank-deficient data
fprintf('\n--- gcPCA rank-deficient data ---\n');
rng(42);
Ra_rd = randn(5, 20);
Rb_rd = randn(5, 20);
try
    [~, ~, X_rd] = gcPCA(Ra_rd, Rb_rd, 4);
    [test_count, pass_count] = check(~isempty(X_rd), 'Rank-deficient data handled', test_count, pass_count);
catch ME
    test_count = test_count + 1;
    fail_count = fail_count + 1;
    failures{end+1} = sprintf('Rank-deficient: %s', ME.message);
    fprintf('  FAIL: Rank-deficient: %s\n', ME.message);
end

%% Test gcPCA: equal data gives zero v4 values
fprintf('\n--- gcPCA equal data ---\n');
rng(42);
data_eq = randn(30, 10);
[~, S_eq, ~] = gcPCA(data_eq, data_eq, 4, 'normalize', false);
objval_eq = S_eq.objval;
[test_count, pass_count] = check(all(abs(objval_eq) < 1e-10), ...
    'Equal data gives zero v4 objective values', test_count, pass_count);

%% Test gcPCA: reproducibility
fprintf('\n--- gcPCA reproducibility ---\n');
[~, ~, X1] = gcPCA(Ra, Rb, 4);
[~, ~, X2] = gcPCA(Ra, Rb, 4);
[test_count, pass_count] = check(max(abs(abs(X1(:)) - abs(X2(:)))) < 1e-10, ...
    'Same data gives same loadings', test_count, pass_count);

%% Test sparse_gcPCA: basic fit for v2-v4
fprintf('\n--- sparse_gcPCA ---\n');

lambdas = [0.1, 0.5, 1.0];

for ver = [2, 3, 4]
    try
        [B, S, X] = sparse_gcPCA(Ra_sm, Rb_sm, ver, 'Nsparse', 2, 'lasso_penalty', lambdas);
        [test_count, pass_count] = check(true, sprintf('sparse_gcPCA v%d fits without error', ver), test_count, pass_count);
        [test_count, pass_count] = check(length(X) == length(lambdas), ...
            sprintf('sparse_gcPCA v%d: correct number of loading sets', ver), test_count, pass_count);
        [test_count, pass_count] = check(size(X{1}, 1) == p_sm, ...
            sprintf('sparse_gcPCA v%d: loadings have correct rows', ver), test_count, pass_count);
        [test_count, pass_count] = check(size(X{1}, 2) == 2, ...
            sprintf('sparse_gcPCA v%d: loadings have Nsparse columns', ver), test_count, pass_count);
    catch ME
        test_count = test_count + 1;
        fail_count = fail_count + 1;
        failures{end+1} = sprintf('sparse_gcPCA v%d: %s', ver, ME.message);
        fprintf('  FAIL: sparse_gcPCA v%d: %s\n', ver, ME.message);
    end
end

%% Test sparse_gcPCA: scores output
fprintf('\n--- sparse_gcPCA scores ---\n');
[B, S, X] = sparse_gcPCA(Ra_sm, Rb_sm, 4, 'Nsparse', 2, 'lasso_penalty', lambdas);
[test_count, pass_count] = check(length(B.a) == length(lambdas), ...
    'B.a cell length matches lambdas', test_count, pass_count);
[test_count, pass_count] = check(length(B.b) == length(lambdas), ...
    'B.b cell length matches lambdas', test_count, pass_count);
[test_count, pass_count] = check(length(S.a) == length(lambdas), ...
    'S.a cell length matches lambdas', test_count, pass_count);
[test_count, pass_count] = check(length(S.b) == length(lambdas), ...
    'S.b cell length matches lambdas', test_count, pass_count);

%% Test sparse_gcPCA v1
fprintf('\n--- sparse_gcPCA v1 ---\n');
try
    [B, S, X] = sparse_gcPCA(Ra_sm, Rb_sm, 1, 'Nsparse', 2, 'lasso_penalty', lambdas);
    [test_count, pass_count] = check(true, 'sparse_gcPCA v1 fits without error', test_count, pass_count);
catch ME
    test_count = test_count + 1;
    fail_count = fail_count + 1;
    failures{end+1} = sprintf('sparse_gcPCA v1: %s', ME.message);
    fprintf('  FAIL: sparse_gcPCA v1: %s\n', ME.message);
end

%% Summary
fprintf('\n========================================\n');
fprintf('  Results: %d passed, %d failed out of %d tests\n', pass_count, fail_count, test_count);
fprintf('========================================\n');

if fail_count > 0
    fprintf('\nFailed tests:\n');
    for i = 1:length(failures)
        fprintf('  - %s\n', failures{i});
    end
    exit(1);
else
    fprintf('\nAll tests passed!\n');
    exit(0);
end

%% Helper function
function [tc, pc] = check(condition, msg, tc, pc)
    tc = tc + 1;
    if condition
        pc = pc + 1;
        fprintf('  PASS: %s\n', msg);
    else
        fprintf('  FAIL: %s\n', msg);
    end
end
